File: core/src/net/oryanmat/rain/Board.java
package net.oryanmat.rain;

import com.badlogic.gdx.graphics.Color;

import static net.oryanmat.rain.TetraRain.COLUMNS;
import static net.oryanmat.rain.TetraRain.ROWS;
import static net.oryanmat.rain.TetraRain.SHAPE_SIZE;

public class Board {
    Block[][] board = new Block[COLUMNS][ROWS + 2];

    public Board() {
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[i].length; j++) {
                board[i][j] = new Block();
            }
        }
    }

    public boolean occupiedAt(int column, int row) {
        return !board[column][row].getEmpty();
    }

    public boolean occupiedOrOutOfBounds(int column, int row) {
        return column >= COLUMNS || column < 0 || row < 0 || occupiedAt(column, row);
    }

    public Color colorAt(int column, int row) {
        return board[column][row].getColor();
    }

    void setInBoard(Shape shape) {
        for (int i = 0; i < SHAPE_SIZE; i++) {
            for (int j = 0; j < SHAPE_SIZE; j++) {
                if (shape.occupiedAt(i, j)) {
                    Block block = board[shape.column + i][shape.row + j];
                    block.setEmpty(false);
                    block.setColor(shape.color);
                }
            }
        }
    }

    int checkForFullRow() {
        int fullRows = 0;

        for (int j = 0; j < ROWS; j++) {
            if (isRowFull(j)) {
                clearFullRow(j);
                dropBoard(j);
                fullRows++;
                // move back one row, so we don't skip j + 1 which is now j
                j--;
            }
        }

        return fullRows;
    }

    int setInBoardAndCheckFullRow(Shape shape) {
        setInBoard(shape);
        return checkForFullRow();
    }

    private boolean isRowFull(int row) {
        boolean full = true;

        for (int column = 0; column < COLUMNS; column++) {
            full &= occupiedAt(column, row);
        }

        return full;
    }

    private void clearFullRow(int row) {
        for (int i = 0; i < COLUMNS; i++) {
            board[i][row].setEmpty(true);
        }
    }

    private void dropBoard(int row) {
        for (int j = row + 1; j < ROWS; j++) {
            for (int i = 0; i < COLUMNS; i++) {
                board[i][j - 1].setEmpty(board[i][j].getEmpty());
                board[i][j - 1].setColor(board[i][j].getColor());
            }
        }
    }
}

File: core/src/net/oryanmat/rain/Board.kt
package net.oryanmat.rain

import net.oryanmat.rain.TetraRain.*

internal class Board {
    private val board = Array(COLUMNS, { Array(ROWS + 2, { Block() }) })

    fun occupiedAt(column: Int, row: Int) = board[column][row].empty.not()

    fun colorAt(column: Int, row: Int) = board[column][row].color

    fun occupiedOrOutOfBounds(column: Int, row: Int): Boolean {
        return column >= COLUMNS || column < 0 || row < 0 || occupiedAt(column, row)
    }

    private fun setInBoard(shape: Shape) {
        for (i in 0..SHAPE_SIZE - 1) {
            for (j in 0..SHAPE_SIZE - 1) {
                if (shape.occupiedAt(i, j)) {
                    val block = board[shape.column + i][shape.row + j]
                    block.empty = false
                    block.color = shape.color
                }
            }
        }
    }

    private fun checkForFullRow(): Int {
        var fullRows = 0

        var j = 0
        while (j < ROWS) {
            if (isRowFull(j)) {
                clearFullRow(j)
                dropBoard(j)
                fullRows++
                // move back one row, so we don't skip j + 1 which is now j
                j--
            }
            j++
        }

        return fullRows
    }

    fun setInBoardAndCheckFullRow(shape: Shape): Int {
        setInBoard(shape)
        return checkForFullRow()
    }

    private fun isRowFull(row: Int): Boolean {
        var full = true

        for (column in 0..COLUMNS - 1) {
            full = full and occupiedAt(column, row)
        }

        return full
    }

    private fun clearFullRow(row: Int) {
        for (i in 0..COLUMNS - 1) {
            board[i][row].empty = true
        }
    }

    private fun dropBoard(row: Int) {
        for (j in row + 1..ROWS - 1) {
            for (i in 0..COLUMNS - 1) {
                board[i][j - 1].empty = board[i][j].empty
                board[i][j - 1].color = board[i][j].color
            }
        }
    }
}
